Enhance Loop Unroller

mostafaelhoushi opened the pull request:
We have noticed that the unrolling factor passed to `opt` as `--unroll-count=<num>` cli option doesn't always take effect.
This PR provides a custom LLVM loop unroller that enforces the unroll count factor provided in an action.. 

To run custom unroller:
```
bazel run //examples/example_unrolling_service/loop_unroller:loop_unroller -- <input>.ll -o <output>.ll  --funroll-count=<num>
```

To run end-to-end example:
```
bazel run //examples/example_unrolling_service:example
```

mostafaelhoushi committed:
start with a LLVM analysis file

mostafaelhoushi committed:
use an IR pass example

mostafaelhoushi committed:
clean up code

mostafaelhoushi committed:
remove header file

mostafaelhoushi committed:
change name

mostafaelhoushi committed:
everything in class declaration

mostafaelhoushi committed:
start couning number of loops

mostafaelhoushi committed:
show 2 methods to count number of loops

mostafaelhoushi committed:
fix name

ChrisCummins commented on examples/example_unrolling_service/loop_unroller/loop_unroller.cc:
```diff
@@ -0,0 +1,147 @@
+#include <algorithm>
```
This file needs a license header

mostafaelhoushi replied:
Done

ChrisCummins commented on examples/example_unrolling_service/loop_unroller/loop_unroller.cc:
```diff
@@ -0,0 +1,147 @@
+#define METHOD 1
+
+#if (METHOD == 1)
+// obtained from https://stackoverflow.com/a/33565910/3880948
```
[LGTM](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/third_party/llvm/InstCount.cc#L13) :)

mostafaelhoushi committed:
fix by Hugh Leather

mostafaelhoushi commented:
Update: Thanks for @hughleat  for his help. Counting the loops is now solved. Next step is to update the loop metadata

ChrisCummins commented on examples/example_unrolling_service/loop_unroller/loop_unroller.cc:
```diff
@@ -0,0 +1,119 @@
+                                    cl::value_desc("filename"), cl::init("-"));
+
+namespace llvm {
+// The INITIALIZE_PASS_XXX macros put the initialiser in the llvm namespace.
```
Total bikeshedding, but what about putting the whole file (except `main`) inside `llvm::`?

mostafaelhoushi replied:
Done

ChrisCummins commented on examples/example_unrolling_service/loop_unroller/loop_unroller.cc:
```diff
@@ -0,0 +1,119 @@
+    LoopInfo& LI = getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
+    auto Loops = LI.getLoopsInPreorder();
+
+    // Should reall account for module, too.
```
Typo "reall"

ChrisCummins reviewed (commented):
I had a pick through the code and it all looks good to me (minor comments inline). I'm guessing with this being WIP that the glue code to interface with an environment is coming later. Just ping me whenever you want me to take a look through :-)

Cheers,
Chris

mostafaelhoushi committed:
move all file into llvm{} and hack meta data change

mostafaelhoushi committed:
add separate loop passes to count and to configure

mostafaelhoushi committed:
add meta data and call loop unroll

mostafaelhoushi committed:
pass unroll enable and count as cli

mostafaelhoushi committed:
call loop unroller in env service

mostafaelhoushi committed:
start a README file

mostafaelhoushi committed:
invoke loop_unroller in our unrolling env example

mostafaelhoushi committed:
add header

mostafaelhoushi commented:
> I had a pick through the code and it all looks good to me (minor comments inline). I'm guessing with this being WIP that the glue code to interface with an environment is coming later. Just ping me whenever you want me to take a look through :-)
> 
> Cheers, Chris

Thanks Chris. I have addressed all the comments.
Yes, now I have updated the environment to call the custom unroller. I have also made it optional in case someone wants to use LLVM's default unroller

mostafaelhoushi committed:
fill README file

mostafaelhoushi committed:
add option to output to bitcode file or IR assembly

mostafaelhoushi committed:
make -S option appear when calling --help

mostafaelhoushi commented:
> Just ping me whenever you want me to take a look through :-)

I merged it without pinging you :( 
Do you want to take another look and make changes?

ChrisCummins commented:
> I merged it without pinging you :( Do you want to take another look and make changes?

No worries! Code looks great :) I'll try it out and let you know if i encounter any problems

Cheers,
Chris
